client/sources/ok_test/sqlite.py

```diff
@@ -0,0 +1,135 @@
+from client.utils import timer
+import importlib
+import os
+
+class SqliteConsole(interpreter.Console):
+    PS1 = 'sqlite> '
+    PS2 = '   ...> '
+
+    MODULE = 'sqlite3'
+    VERSION = (3, 8, 3)
+    _output_fn = str
+
+    def __init__(self, verbose, interactive, timeout=None):
+        """Loads the Scheme module from the current working directory
+        before calling the superclass constructor.
+        """
+        super().__init__(verbose, interactive, timeout)
+
+    def load(self, code, setup='', teardown=''):
+        """Prepares a set of setup, test, and teardown code to be
+        run in the console.
+        """
+        self.sqlite3 = self._import_sqlite()
+        super().load(code, setup, teardown)
+        self._conn = self.sqlite3.connect(':memory:', check_same_thread=False)
+
+    def interact(self):
+        """Opens up an interactive session with the current state of
+        the console.
+        """
+        # TODO(albert)
+
+    def evaluate(self, code):
+        if code.startswith('.'):
+            try:
+                self.evaluate_dot(code)
+            except interpreter.ConsoleException as e:
+                print('Error: {}'.format(e))
+                raise
+            return
+        try:
+            cursor = timer.timed(self.timeout, self._conn.execute, (code,))
+        except exceptions.Timeout as e:
+            print('Error: evaluation exceeded {} seconds.'.format(e.timeout))
+            raise interpreter.ConsoleException(e)
+        except self.sqlite3.Error as e:
+            print('Error: {}'.format(e))
+            raise interpreter.ConsoleException(e, exception_type='Error')
+        else:
+            return cursor
+
+    def _compare(self, expected, code):
+        try:
+            cursor = self.evaluate(code)
+        except interpreter.ConsoleException as e:
+            actual = [e.exception_type]
+        else:
+            actual = self.format_rows(cursor)
+
+        if expected != 'Error':
+            expected = set(expected.split('\n'))
+        if expected != actual:
+            print()
+            print('# Error: expected')
+            print('\n'.join('#     {}'.format(line)
+                            for line in expected))
+            print('# but got')
+            print('\n'.join('#     {}'.format(line)
+                            for line in actual))
+            raise interpreter.ConsoleException
+
+    def _import_sqlite(self):
+        try:
+            sqlite = importlib.import_module(self.MODULE)
+        except ImportError:
+            print()
+            raise exceptions.ProtocolException(
+                    'Could not import sqlite3. '
+                    'Make sure you have installed sqlite3')
+        if sqlite.sqlite_version_info < self.VERSION:
+            raise exceptions.ProtocolException(
+            'You are running an outdated version of sqlite3:\n'
+            '    {}\n'
+            'Please install sqlite version {} '
```

Comment:
It would be nice if we gave them a link, but I don't really care very much about this since I don't think that we should be keeping track of when a link gets outdated.